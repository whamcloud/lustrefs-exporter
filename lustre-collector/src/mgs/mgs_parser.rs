// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{
    base_parsers::{digits, param, period, target},
    stats_parser::stats,
    types::{Param, Record, Stat, Target, TargetStat, TargetStats, TargetVariant},
};
use combine::{
    attempt, choice,
    error::ParseError,
    parser::char::{newline, string},
    stream::Stream,
    Parser,
};

pub const STATS: &str = "stats";
pub const THREADS_MIN: &str = "threads_min";
pub const THREADS_MAX: &str = "threads_max";
pub const THREADS_STARTED: &str = "threads_started";
pub const NUM_EXPORTS: &str = "num_exports";

pub fn params() -> Vec<String> {
    [
        format!("mgs.*.mgs.{STATS}"),
        format!("mgs.*.mgs.{THREADS_MAX}"),
        format!("mgs.*.mgs.{THREADS_MIN}"),
        format!("mgs.*.mgs.{THREADS_STARTED}"),
        format!("mgs.*.{NUM_EXPORTS}"),
    ]
    .iter()
    .map(|x| x.to_owned())
    .collect::<Vec<_>>()
}

#[derive(Debug)]
enum MgsStat {
    Stats(Vec<Stat>),
    ThreadsMin(u64),
    ThreadsMax(u64),
    ThreadsStarted(u64),
    NumExports(u64),
}

/// Parses the name of a target
fn target_name<I>() -> impl Parser<I, Output = Target>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    (
        attempt(string("mgs")).skip(period()),
        target().skip(period()),
    )
        .map(|(_, x)| x)
        .message("while parsing target_name")
}

fn mgs_stat<I>() -> impl Parser<I, Output = (Param, MgsStat)>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    choice((
        (
            param(NUM_EXPORTS),
            digits().skip(newline()).map(MgsStat::NumExports),
        ),
        (
            string("mgs").skip(period()),
            choice((
                (param(STATS), stats().map(MgsStat::Stats)),
                (
                    param(THREADS_MIN),
                    digits().skip(newline()).map(MgsStat::ThreadsMin),
                ),
                (
                    param(THREADS_MAX),
                    digits().skip(newline()).map(MgsStat::ThreadsMax),
                ),
                (
                    param(THREADS_STARTED),
                    digits().skip(newline()).map(MgsStat::ThreadsStarted),
                ),
            )),
        )
            .map(|(_, (y, z))| (y, z)),
    ))
    .message("while parsing mgs stats")
}

pub fn parse<I>() -> impl Parser<I, Output = Record>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    (target_name(), mgs_stat())
        .map(|(target, (param, value))| match value {
            MgsStat::Stats(value) => TargetStats::Stats(TargetStat {
                kind: TargetVariant::Mgt,
                target,
                param,
                value,
            }),
            MgsStat::NumExports(value) => TargetStats::NumExports(TargetStat {
                kind: TargetVariant::Mgt,
                target,
                param,
                value,
            }),
            MgsStat::ThreadsMin(value) => TargetStats::ThreadsMin(TargetStat {
                kind: TargetVariant::Mgt,
                target,
                param,
                value,
            }),
            MgsStat::ThreadsMax(value) => TargetStats::ThreadsMax(TargetStat {
                kind: TargetVariant::Mgt,
                target,
                param,
                value,
            }),
            MgsStat::ThreadsStarted(value) => TargetStats::ThreadsStarted(TargetStat {
                kind: TargetVariant::Mgt,
                target,
                param,
                value,
            }),
        })
        .map(Record::Target)
        .message("while parsing mgs params")
}

#[cfg(test)]
mod tests {
    use super::*;
    use combine::{many, parser::EasyParser};
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse() {
        let x = r#"mgs.MGS.mgs.stats=
snapshot_time             1596728874.484750908 secs.nsecs
req_waittime              31280 samples [usec] 11 2695 5020274 1032267156
req_qdepth                31280 samples [reqs] 0 1 56 56
req_active                31280 samples [reqs] 1 2 36625 47315
req_timeout               31280 samples [sec] 1 10 31289 31379
reqbuf_avail              85192 samples [bufs] 62 64 5364658 337866142
ldlm_plain_enqueue        201 samples [reqs] 1 1 201 201
mgs_connect               9 samples [usec] 52 5165 19362 66639088
mgs_disconnect            4 samples [usec] 50 92 265 18709
mgs_target_reg            90 samples [usec] 874 163383 1262544 91852108168
mgs_config_read           41 samples [usec] 41 2203 26823 32448779
obd_ping                  30339 samples [usec] 3 4398 1552005 134387261
llog_origin_handle_open   153 samples [usec] 29 16443 25516 270992222
llog_origin_handle_next_block 298 samples [usec] 24 31952 141030 2788155300
llog_origin_handle_read_header 145 samples [usec] 25 44125 192095 4905765639
mgs.MGS.mgs.threads_max=32
mgs.MGS.mgs.threads_min=3
mgs.MGS.mgs.threads_started=4
mgs.MGS.num_exports=5
"#;

        let result: (Vec<_>, _) = many(parse()).easy_parse(x).unwrap();

        assert_debug_snapshot!(result)
    }
}
