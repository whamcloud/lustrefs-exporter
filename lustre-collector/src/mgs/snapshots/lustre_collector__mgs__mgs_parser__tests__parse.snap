---
source: src/mgs/mgs_parser.rs
expression: result

---
(
    [
        Target(
            Stats(
                TargetStat {
                    kind: Mgt,
                    param: Param(
                        "stats",
                    ),
                    target: Target(
                        "MGS",
                    ),
                    value: [
                        Stat {
                            name: "req_waittime",
                            units: "usec",
                            samples: 31280,
                            min: Some(
                                11,
                            ),
                            max: Some(
                                2695,
                            ),
                            sum: Some(
                                5020274,
                            ),
                            sumsquare: Some(
                                1032267156,
                            ),
                        },
                        Stat {
                            name: "req_qdepth",
                            units: "reqs",
                            samples: 31280,
                            min: Some(
                                0,
                            ),
                            max: Some(
                                1,
                            ),
                            sum: Some(
                                56,
                            ),
                            sumsquare: Some(
                                56,
                            ),
                        },
                        Stat {
                            name: "req_active",
                            units: "reqs",
                            samples: 31280,
                            min: Some(
                                1,
                            ),
                            max: Some(
                                2,
                            ),
                            sum: Some(
                                36625,
                            ),
                            sumsquare: Some(
                                47315,
                            ),
                        },
                        Stat {
                            name: "req_timeout",
                            units: "sec",
                            samples: 31280,
                            min: Some(
                                1,
                            ),
                            max: Some(
                                10,
                            ),
                            sum: Some(
                                31289,
                            ),
                            sumsquare: Some(
                                31379,
                            ),
                        },
                        Stat {
                            name: "reqbuf_avail",
                            units: "bufs",
                            samples: 85192,
                            min: Some(
                                62,
                            ),
                            max: Some(
                                64,
                            ),
                            sum: Some(
                                5364658,
                            ),
                            sumsquare: Some(
                                337866142,
                            ),
                        },
                        Stat {
                            name: "ldlm_plain_enqueue",
                            units: "reqs",
                            samples: 201,
                            min: Some(
                                1,
                            ),
                            max: Some(
                                1,
                            ),
                            sum: Some(
                                201,
                            ),
                            sumsquare: Some(
                                201,
                            ),
                        },
                        Stat {
                            name: "mgs_connect",
                            units: "usec",
                            samples: 9,
                            min: Some(
                                52,
                            ),
                            max: Some(
                                5165,
                            ),
                            sum: Some(
                                19362,
                            ),
                            sumsquare: Some(
                                66639088,
                            ),
                        },
                        Stat {
                            name: "mgs_disconnect",
                            units: "usec",
                            samples: 4,
                            min: Some(
                                50,
                            ),
                            max: Some(
                                92,
                            ),
                            sum: Some(
                                265,
                            ),
                            sumsquare: Some(
                                18709,
                            ),
                        },
                        Stat {
                            name: "mgs_target_reg",
                            units: "usec",
                            samples: 90,
                            min: Some(
                                874,
                            ),
                            max: Some(
                                163383,
                            ),
                            sum: Some(
                                1262544,
                            ),
                            sumsquare: Some(
                                91852108168,
                            ),
                        },
                        Stat {
                            name: "mgs_config_read",
                            units: "usec",
                            samples: 41,
                            min: Some(
                                41,
                            ),
                            max: Some(
                                2203,
                            ),
                            sum: Some(
                                26823,
                            ),
                            sumsquare: Some(
                                32448779,
                            ),
                        },
                        Stat {
                            name: "obd_ping",
                            units: "usec",
                            samples: 30339,
                            min: Some(
                                3,
                            ),
                            max: Some(
                                4398,
                            ),
                            sum: Some(
                                1552005,
                            ),
                            sumsquare: Some(
                                134387261,
                            ),
                        },
                        Stat {
                            name: "llog_origin_handle_open",
                            units: "usec",
                            samples: 153,
                            min: Some(
                                29,
                            ),
                            max: Some(
                                16443,
                            ),
                            sum: Some(
                                25516,
                            ),
                            sumsquare: Some(
                                270992222,
                            ),
                        },
                        Stat {
                            name: "llog_origin_handle_next_block",
                            units: "usec",
                            samples: 298,
                            min: Some(
                                24,
                            ),
                            max: Some(
                                31952,
                            ),
                            sum: Some(
                                141030,
                            ),
                            sumsquare: Some(
                                2788155300,
                            ),
                        },
                        Stat {
                            name: "llog_origin_handle_read_header",
                            units: "usec",
                            samples: 145,
                            min: Some(
                                25,
                            ),
                            max: Some(
                                44125,
                            ),
                            sum: Some(
                                192095,
                            ),
                            sumsquare: Some(
                                4905765639,
                            ),
                        },
                    ],
                },
            ),
        ),
        Target(
            ThreadsMax(
                TargetStat {
                    kind: Mgt,
                    param: Param(
                        "threads_max",
                    ),
                    target: Target(
                        "MGS",
                    ),
                    value: 32,
                },
            ),
        ),
        Target(
            ThreadsMin(
                TargetStat {
                    kind: Mgt,
                    param: Param(
                        "threads_min",
                    ),
                    target: Target(
                        "MGS",
                    ),
                    value: 3,
                },
            ),
        ),
        Target(
            ThreadsStarted(
                TargetStat {
                    kind: Mgt,
                    param: Param(
                        "threads_started",
                    ),
                    target: Target(
                        "MGS",
                    ),
                    value: 4,
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: Mgt,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "MGS",
                    ),
                    value: 5,
                },
            ),
        ),
    ],
    "",
)
